### ISSUE TITLE
(Optionally) "steal" keyboard input in panel mode?

### ISSUE THREAD

**Issue by josswright** [kovidgoyal/kitty#2195]
I just discovered the wonderful option for `kitty` to run as a dock.

I was desperately hoping that this would let me use kitty as a handy program launcher under `i3`. (I currently use a normal `kitty` window for this, but it's annoying to have the full extra window pop up.) I was hoping for something that felt like http://guake-project.org/ , but that nicely shifted up my `i3` windows.

I was planning on something like this:

    kitty +kitten panel --edge bottom --lines 3 -o allow_remote_control=yes zsh

The problem, of course, is that `i3` does not allow a dock window to have focus, and so I can't type in the terminal that appears. I've tried to use the remote control options to set focus, but I'm not having any success. (Maybe I'm doing it wrong, but I still can't seem to focus that window.)

Is there any way that kitty could override `i3` and 'steal' X11 input focus -- probably globally -- for a temporary window like this? (Or even to have a remote command that allows a 'hard' form of the focus-window command?) 

I can't tell if this would be best addressed by `i3` allowing a given dock window to be focused, or by allowing `kitty` to steal focus for the entire X session. It feels more natural for `kitty` to do it, and I suspect that I won't have much success convincing `i3` to add the feature for a dock window to get focus.

**Comment by kovidgoyal** [comment]
I dont think GLFW (the toolkit kitty uses exposes XGrabKeyboard). But
give the global nature of X it should be easy to do. Use a script like
https://www.perlmonks.org/?node_id=693422

to grab the keyboard using XGrabKeyboard. Run that script instead of zsh
and in it grab keyboard, start shell, wait for shell to exit, ungrab
keyboard.

Personally, I use floating rather than docked windows for this kind of
thing. I dont like all windows being resized when launching programs.

**Comment by kovidgoyal** [comment]
Oh and kitty sets an environment variable WINDOWID which you can use to get the window id of the window the script is running in.

**Comment by josswright** [comment]
Thanks for the tip! I also used to use a floating window for this, but I've been looking for a dock-style solution for a while now. I'll report back if I get this working.

**Comment by josswright** [comment]
> Thanks for the tip! I also used to use a floating window for this, but I've been looking for a dock-style solution for a while now. I'll report back if I get this working.

After a morning playing with tihs, and looking at the various libraries and options for grabbing input, it looks like the dock window probably isn't worth the effort. In case anyone else is looking for a similar approach, I'll describe my not-quite-satisfactory solution here.

I wrote a script that uses i3-msg to focus the top-level container, make it a vertical split, call my terminal command, then resize that new window to 10% of the window height. It's not quite as smooth as the appearance of a dock window, but it avoids me having to write a keygrabber that passes input through to the dock window. (Something that was far more complicated than I expected, as the listener doesn't typically pass things through to the process that called it.)

The script I'm using now looks like this:

    i3-msg 'focus parent'
    i3-msg 'focus parent'
    i3-msg 'focus parent'
    i3-msg 'split v' 
    kitty &!
    sleep 0.4
    i3-msg 'resize set 100ppt 10ppt'

It feels hacky, assumes that you don't have more than three layers of nested windows on screen, and relies on a 0.4 second sleep to let the window appear on screen before resizing. It's more or less working for me, though. If someone more skilled can work out a script to get the dock window to grab input and pass it seamlessly through to zsh, I'd love to hear about it!

**Comment by kovidgoyal** [comment]
Does calling XGrabKeyboard with WINDOWID of the dock window not pass key
events to the dock window?

**Comment by josswright** [comment]
> Does calling XGrabKeyboard with WINDOWID of the dock window not pass key
> events to the dock window?

The script that you linked will take the focus so that keyboard events are grabbed, but they don't seem to be passed to the zsh process directly. The events seemed to need to be handled, or passed through, explicitly. Most of the similar scripts that I could find were based around being global hotkey handlers, so were focused on explicit hooks to handle keybindings, not just a pass-through to another process. I had a look into it, but it started to look more complicated than I had time to manage. 

I'm sure that there is a way to do it, it's just that going from my zero knowledge of Xlib and this kind of event handling to making this work was more than I could manage in a reasonable time frame!

**Comment by josswright** [comment]
(I'd love to figure this out, actually, as this would be incredibly helpful for `dmenu` as well.)

**Comment by kovidgoyal** [comment]
On Sat, Dec 14, 2019 at 07:21:59AM -0800, Joss Wright wrote:
> > Does calling XGrabKeyboard with WINDOWID of the dock window not pass key
> > events to the dock window?
> 
> The script that you linked will take the focus so that keyboard events are grabbed, but they don't seem to be passed to the zsh process directly. The events seemed to need to be handled, or passed through, explicitly. Most of the similar scripts that I could find were based around being global hotkey handlers, so were focused on explicit hooks to handle keybindings, not just a pass-through to another process. I had a look into it, but it started to look more complicated than I had time to manage. 

A script cannot take focus, it does not have an X11 window associated
with it. What I would have thought would happen is the window identified
by WINDOWID which is the kitty window the script is running in should
get keyboard focus and handle keyboard events normally.

**Comment by josswright** [comment]
> On Sat, Dec 14, 2019 at 07:21:59AM -0800, Joss Wright wrote: > Does calling XGrabKeyboard with WINDOWID of the dock window not pass key > events to the dock window? The script that you linked will take the focus so that keyboard events are grabbed, but they don't seem to be passed to the zsh process directly. The events seemed to need to be handled, or passed through, explicitly. Most of the similar scripts that I could find were based around being global hotkey handlers, so were focused on explicit hooks to handle keybindings, not just a pass-through to another process. I had a look into it, but it started to look more complicated than I had time to manage.
> A script cannot take focus, it does not have an X11 window associated with it. What I would have thought would happen is the window identified by WINDOWID which is the kitty window the script is running in should get keyboard focus and handle keyboard events normally.

The script that calls the XGrabKeyboard, running in an X client. (Note that the window ID in the call should be the root window, as we're trying to grab global keyboard events from the X server.)

The problem is that that process, which is running in the X client, seems to grab and hold the input, and not pass it on to any subprocess. A good example is the python script in this question: https://stackoverflow.com/questions/18160990/grabbing-keyboard-doesnt-allow-changing-focus

I've saved that script to `grabby.py`, and run it as follows:

    kitty +kitten panel --lines 4 python tmp/grabby.py

It will run, and the keyboard events will be handled, with the event types output in kitty. If I replace the while loop with a simple `subprocess.run(["zsh"])`, though, it just sits there until the process naturally exits. The incoming events grabbed by the call need to be handled explicitly, from what I can see. (This is where I'm getting to the very end of my limited knowledge of X input handling.)

I'm not sure exactly why this happens, or how to solve it. I hope that I'm missing something obvious!

**Comment by kovidgoyal** [comment]
Sounds like you can make that happen by grabbing the keyboard in the
panel kitten. Try it by creating a custom kitten, just copy the panel
kitten code, and in setup_x11_window try grabbing the keyboard. You can
get the Display pointer with x11_display() (see utils.py for an
example)

https://sw.kovidgoyal.net/kitty/kittens/custom.html 

On Sat, Dec 14, 2019 at 12:55:38PM -0800, Joss Wright wrote:
> > On Sat, Dec 14, 2019 at 07:21:59AM -0800, Joss Wright wrote: > Does calling XGrabKeyboard with WINDOWID of the dock window not pass key > events to the dock window? The script that you linked will take the focus so that keyboard events are grabbed, but they don't seem to be passed to the zsh process directly. The events seemed to need to be handled, or passed through, explicitly. Most of the similar scripts that I could find were based around being global hotkey handlers, so were focused on explicit hooks to handle keybindings, not just a pass-through to another process. I had a look into it, but it started to look more complicated than I had time to manage.
> > A script cannot take focus, it does not have an X11 window associated with it. What I would have thought would happen is the window identified by WINDOWID which is the kitty window the script is running in should get keyboard focus and handle keyboard events normally.
> 
> The script that calls the XGrabKeyboard, running in an X client. (Note that the window ID in the call should be the root window, as we're trying to grab global keyboard events from the X server.)
> 
> The problem is that that process, which is running in the X client, seems to grab and hold the input, and not pass it on to any subprocess. A good example is the python script in this question: https://stackoverflow.com/questions/18160990/grabbing-keyboard-doesnt-allow-changing-focus
> 
> I've saved that script to `grabby.py`, and run it as follows:
> 
>     kitty +kitten panel --lines 4 python tmp/grabby.py
> 
> It will run, and the keyboard events will be handled, with the event types output in kitty. If I replace the while loop with a simple `subprocess.run(["zsh"])`, though, it just sits there until the process naturally exits. The incoming events grabbed by the call need to be handled explicitly, from what I can see. (This is where I'm getting to the very end of my limited knowledge of X input handling.)
> 
> I'm not sure exactly why this happens, or how to solve it. I hope that I'm missing something obvious!
> 
> -- 
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/issues/2195#issuecomment-565752465

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by kovidgoyal** [comment]
Oh and if you really want to get hacky, you can have two windows, a kitty panel that runs a kitty floating window placed on top of the panel.

**Comment by josswright** [comment]
I suspect that that might be a hideous, but more workable, solution.

As far as I can tell, the grabbing process needs to run a loop that manually passes each key event through to the client. I can make a custom kitten that reproduces the event monitoring code, but you can't seem just to grab focus and simply go on to other things. 

I'm surprised that I can't find any code examples of people doing this, to be honest.

**Comment by kovidgoyal** [comment]
On Sun, Dec 15, 2019 at 03:09:01AM -0800, Joss Wright wrote:
> I suspect that that might be a hideous, but more workable, solution.
> 
> As far as I can tell, the grabbing process needs to run a loop that manually passes each key event through to the client. I can make a custom kitten that reproduces the event monitoring code, but you can't seem just to grab focus and simply go on to other things. 

So if you do what I suggest, then you are calling XGrabKeyboard in the
kitty process that has created the panel window,
which processes events itself, so you should not need to process events
or pass them anywhere.

**Comment by josswright** [comment]
> On Sun, Dec 15, 2019 at 03:09:01AM -0800, Joss Wright wrote: I suspect that that might be a hideous, but more workable, solution. As far as I can tell, the grabbing process needs to run a loop that manually passes each key event through to the client. I can make a custom kitten that reproduces the event monitoring code, but you can't seem just to grab focus and simply go on to other things.
> So if you do what I suggest, then you are calling XGrabKeyboard in the kitty process that has created the panel window, which processes events itself, so you should not need to process events or pass them anywhere.

At the end of `setup_x11_window` in a clone of the panel kitten, I now have:

    # current display
    global display,root
    display = Display()
    root = display.screen().root

    root.change_attributes(event_mask = X.KeyPressMask|X.KeyReleaseMask)
    root.grab_keyboard(True, X.GrabModeAsync, X.GrabModeAsync,X.CurrentTime)

The dock window appears, with zsh running. The keyboard is grabbed, but nothing shows up in the terminal when I type. (As I haven't ungrabbed the keyboard at the end, I have to drop out to the console and kill the kitty process to get the keyboard back.)

I've tried various options, and have been looking into both XGrabKey, which seems to have slightly different behaviour, and XSetInputFocus in case that might work. (I'm fairly sure at this point that XSetInputFocus doesn't.)

**Comment by kovidgoyal** [comment]
What happens if you use the window identified by the window id passed into the setup_x11_window() function instead of the root window? Also try passing both true and false for owner_events in XGrabKeyboard

**Comment by josswright** [comment]
Same behaviour for all those options (with owner events set to True or False).

    # current display
    global display,root
    display = Display()
    root = display.create_resource_object('window', int( os.environ['WINDOWID'] ))

    root.change_attributes(event_mask = X.KeyPressMask|X.KeyReleaseMask)
    root.grab_keyboard(True, X.GrabModeAsync, X.GrabModeAsync,X.CurrentTime)

**Comment by josswright** [comment]
Notably, though, it's not the focus that's an issue, it's the keypress handling. As I said, when I include the event loop to read events, the window can respond to that and display the event types. 

I've been looking at ReplayKeyboard in XAllowEvents, and that's what led me to the XGrabKey rather than XGrabKeyboard, which seems to be somehow passive rather than active. The documentation around this area seems very arcane, though. I also don't know if I can use XGrabKey to grab everything.

**Comment by kovidgoyal** [comment]
Ugh, its probably just easier to use the hack I described or patch i3 to not prevent keyboard focus on dock windows with a specific class.

**Comment by josswright** [comment]
I agree!

**Comment by josswright** [comment]
Success! 

This can be done quite neatly with XSetInputFocus (or set_input_focus from python-xlib). My earlier failures with this were due to it also requiring a flush() call to make the set_input_focus call take effect. (There's an event queue meaning that the request doesn't seem to take effect until you make the next request. I may be misunderstanding, as the documentation is arcane, but I noticed this when it worked after I called a get_input_focus() to check what was going on.)

The following script, run from within the docked kitty window, will get input focus under i3:

    from Xlib import X
    from Xlib.display import Display
    import os
    
    # Get display
    display = Display()
    
    # Create window object from current WINDOWID
    win = display.create_resource_object( 'window', int( os.environ['WINDOWID'] ) )
    
    # Set input focus, then flush event queue.
    display.set_input_focus( win, X.RevertToNone, X.CurrentTime )
    display.flush()

This doesn't work, though, if I put it in `setup_x11_window()` in a kitten. For some reason it seemed to hang when I tried it, so I've moved it to an external script to be called after the window has fully appeared.

As such, my best solution now is to save that script as, say, `$HOME/bin/kittygrab.py`, then have a shell script `kittydock.sh` that simply calls it, then calls zsh.

With those two in place, I can now run:

    kitty +kitten panel --lines 3 --edge bottom bin/kittydock.sh

... and have my ideal docked program launcher in i3! I can't tell you how satisfying this is after a number of years of wishing that this was possible. Thank you!

(I've also started a conversation over on the i3 github about allowing docks to receive keyboard input directly in certain circumstances: https://github.com/i3/i3/issues/3881 , which would allow for a much more robust and general means of doing this.)

**Comment by josswright** [comment]
(The one down side, of course, is that if that window loses focus then there's no simple way either to get it back, or even kill it without something like `xkill`. I'm very happy with it for now, though.)

**Comment by kovidgoyal** [comment]
Cool glad you figured it out :)

**Comment by gandalf3** [comment]
Hello, two and a half years later I've stumbled across this issue while trying to do the exact same thing. I've managed to get it working after working around two main snags:
* The window must be mapped (i.e. visible) before requesting a grab.
  - For the window to be actually mapped immediately when we call XMapWindow, it must be marked *Override Redirect*.
* I wasn't able to get kitty to process the X11 events after the grab unless I executed the grab using the same display connection as kitty itself; I was able to accomplish this via `kitty.fast_data_types.x11_display`.

Here's a gist of a minimal popup kitten, which at least works for me: https://gist.github.com/gandalf3/67285f8e4d5e7f740890889f230738a9

If anyone else is still interested in this, I hope this helps!
